Fix Jetpack carousel data enrichment

daledupreez opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes Automattic/wp-calypso#64638
Fixes #22726

#### Changes proposed in this Pull Request:

* This PR changes the way we add additional data attributes for carousel-enabled gallery blocks so that we can correctly add the data for reusable blocks.
* The existing code was using `'the_content'` filter to add the `data-carousel-extra` attribute to the outermost gallery element, but this meant that it was not being run for reusable blocks, because the main post content won't contain the resolved block content yet
* The new code uses the [`render_block_{name}`](https://developer.wordpress.org/reference/hooks/render_block_this-name/) filter to add the data attributes to the first HTML element for the gallery - this now works for reusable and standard blocks

##### Possible compatibility issues

1. We now add the data attributes to the first HTML element, which is different to the previous code, which might add it to up to 4 elements in the gallery tree. Nobody _should_ have been relying on this behaviour, but it was still there. Furthermore, we now perform this addition for each block independently instead of once for all the content.
2. We call the `jp_carousel_add_data_to_container` filter for every gallery block, instead of once for all the content. This may be unexpected for some use cases.
3. We've removed the `check_content_for_blocks()` function from the `Jetpack_Carousel` class. This was marked as public, so may have been used by someone.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

This is a bugfix rather than a product change.

#### Does this pull request change what data or activity we track or use?

The PR doesn't affect what data we record or activity we track.

#### Testing instructions:

* Create a test site, and ensure that it has the Jetpack and Jetpack Beta plugins installed
* Ensure that the Carousel feature is enabled via the Media section of the _Writing_ tab in Jetpack Settings
* Set up the test site so that you have four separate posts with each post containing exactly one of the following blocks:
  - A standard Gallery block with at least two images
  - A standard Tiled Gallery block with at least two images
  - A reusable block that contains a Gallery block with at least two images
  - A reusable block that contains a Tiled Gallery block with at least two images
* Open each of the four posts above in new tabs
* Verify that the two base galleries open the carousel when an image in the gallery is clicked on, and the two galleries within reusable blocks do _not_ open the carousel when any images are clicked on
  - You can also take a look at the markup to see the cases where we have a profusion of `data-carousel-extra` attributes in the element tree
* Use the Jetpack Beta tool to switch to the code for this branch (`update/jetpack-gallery-data-enrichment`)
* Refresh each of the four pages, and verify that all four galleries open the carousel when you click on an image
* Inspect each of the gallery elements, and verify that only one `data-carousel-extra` attributes exists for each gallery element

##### Regression tests

We also want to make sure we're not introducing any regressions, as noted in [this comment](https://github.com/Automattic/jetpack/pull/25441#issuecomment-1214684123). To that end, verify that all of the following cases continue to work as expected:
* A post with a Classic block that contains a standard gallery
  - After adding a Classic block, you can create a standard gallery by clicking on "Add Media" in the toolbar, selecting multiple images, and then selecting "Thumbnail Gallery" in the _Layout_ dropdown
* A post with a Classic block that contains a tiled gallery
  - After adding a Classic block, you can create a tiled gallery by clicking on "Add Media" in the toolbar, selecting multiple images, and then specifying a "Tiled Mosaic" in the _Layout_ dropdown
* A post with a Classic block that contains single image
  - After adding a Classic block, click on "Add Media" and select a single image
* A post with an Image block with the "Lightbox" toggle enabled
* A post containing all four of the above blocks as well as a gallery block and a tiled gallery block
 - This should ensure we don't get interference between the block and shortcode processing

In all of the cases above, except for the single image inside the classic block, clicking on the image should show a carousel (or a lightbox for the single image block).

daledupreez committed:
Move Jetpack gallery data enrichment to block render filter

daledupreez committed:
changelog

mpkelly commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -284,27 +291,74 @@ class_exists( 'Jetpack_AMP_Support' )
+
+		// Add extra attributes to first HTML element (which may have leading whitespace)
+		return preg_replace(
+			'/^(\s*<(div|ul|figure))/',
```
Can you confirm in what scenario there would be an `ul` root element? I can see `div` is used for shortcodes and `figure` seems to be used for two gallery blocks. Maybe there is a legacy gallery using `ul`?

daledupreez replied:
TBH, this is mostly something that in the existing code -- I wanted to keep it rather than assume whether it could or couldn't appear.

mpkelly replied:
ok, fair enough!

mpkelly reviewed (dismissed):
This has been tested. In addition to the proposed test plan, I have also checked:
- [x] commenting on images when in carousel mode
- [x] a post with a shortcode gallery + block gallery (as there are now two code paths) 

Everything looks good. I am just not sure why the regex needs to include `ul` (see comment).

jeherve commented:
> Set up the test site so that you have four separate posts with each post containing exactly one of the following blocks

I would recommend testing a few scenarios without blocks as well, to make sure we do not introduce any regressions:

- A classic gallery in a classic editor block
- A tiled gallery in a classic editor block
- A single image in a classic editor block

Catching every single one of those use-cases has proven tricky in the past, so there are a few related issues already opened, some of them may even get closed by your PR?

- #22726 
- #25015
- #13531

daledupreez commented:
Thanks for the review and additional context, @jeherve!

I've already expanded the testing instructions to check for regressions that affect the situations you mentioned. I think the fact that I've split up the processing for the shortcode and the blocks means that there are some interesting side effects:
* The gallery block (post)processing is less invasive and overreaching, because it only modifies HTML specific to the block, which also means that it is less likely to impact or interact with the shortcode processing
* At some stage we may be able to tackle a similar refactor for the slideshow block (which may not have a shortcode equivalent)
* This PR doesn't reduce the number of attributes added to the DOM by the shortcode processing, but it means that only occurs for the shortcode, and not for blocks, which should reduce our payload size for all posts that contain gallery blocks

In terms of the other issues you referenced, I managed to work out the following:
 * This PR should fix https://github.com/Automattic/jetpack/issues/22726
 * This PR won't fix https://github.com/Automattic/jetpack/issues/13531, as that is an issue with the code that is now used for shortcodes only -- I've reduced the scope of the problem in this PR, but not eliminated it
 * I've created https://github.com/Automattic/jetpack/pull/25655 to address https://github.com/Automattic/jetpack/issues/25015 -- the problem there is that we were _storing_ the resized URL as the `href` in the markup

mpkelly commented:
Have retested the new Test Plan. Everything looks ok. The Lightbox toggle is well hidden! Luckily it's[ easy to find on Google](https://jetpack.com/support/carousel/#open-a-single-image-in-a-lightbox).

coder-karen committed:
Merge branch 'trunk' into update/jetpack-gallery-data-enrichment

coder-karen reviewed (dismissed):
Nice - all testing scenarios work as they're expected to. I often forget that on a non-WoA Jetpack connected site Site Accelerator needs to be enabled for tiled galleries to work with Classic Blocks or using the Classic Editor - but that worked as expected when enabled.

sdixon194 committed:
Merge remote-tracking branch 'origin/trunk' into update/jetpack-gallery-data-enrichment

daledupreez committed:
Merge remote-tracking branch 'origin/trunk' into update/jetpack-gallery-data-enrichment

coder-karen committed:
Merge branch 'trunk' into update/jetpack-gallery-data-enrichment

coder-karen reviewed (changes requested):
I originally missed that there is a completely deleted function (which I realise was also mentioned in the PR description). 

There is a little more info on deleting a function in Jetpack here: https://github.com/Automattic/jetpack/blob/trunk/docs/coding-guidelines.md#deprecating-code. There are some examples of functions being deprecated here for comparison: https://github.com/Automattic/jetpack/pull/25315/files

daledupreez committed:
Merge branch 'trunk' into update/jetpack-gallery-data-enrichment

daledupreez committed:
Correctly deprecate Jetpack_Carousel->check_content_for_blocks()

daledupreez commented:
@coder-karen, would you mind taking another look? I've re-instated the deleted function, but I have also added `@deprecated $$next-release$$` as an annotation and added a `_deprecated_function()` call to the code (all in 8ab357c51e15f832a903306db8ae55b6db83efff).

coder-karen commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -304,9 +314,81 @@ class_exists( 'Jetpack_AMP_Support' )
+	 * @param array  $block         The parsed block details for the block.
+	 * @return string The fully-processed HTML for the carousel or gallery block.
+	 *
+	 * @since $$next-release$$
```
```suggestion
	 * @since $$next-version$$
```

coder-karen replied:
Same as with the other comment.

coder-karen commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -287,12 +294,15 @@ class_exists( 'Jetpack_AMP_Support' )
 	 * Check if the content of a post uses gallery blocks. To be used by 'the_content' filter.
 	 *
 	 * @since 6.8.0
+	 * @deprecated $$next-release$$ We now hook into the 'block_render_{block_name}' hook to add markup.
```
```suggestion
	 * @deprecated $$next-version$$ We now hook into the 'block_render_{block_name}' hook to add markup.
```

coder-karen replied:
For these, I'm not sure that `$$next-release$$` would work ( I don't see any documentation or use of this in Jetpack currently) - I'm only aware of [`$$next-version$$` ](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/README.md#package-version-annotations).

daledupreez replied:
This was 100% due to me mis-remembering the specific syntax -- I correctly remembered most of it but clearly got the `version` part of it wrong! I've fixed the usages and injected it in the `_deprecated_function()` call as well.

coder-karen reviewed (changes requested):
Left a couple of comments/suggestions just related to the usage of `$$next-release$$`, otherwise the deprecation functionality looks good here.

daledupreez committed:
Merge branch 'trunk' into update/jetpack-gallery-data-enrichment

coder-karen committed:
Merge branch 'trunk' into update/jetpack-gallery-data-enrichment

daledupreez committed:
Fix next-version syntax

coder-karen reviewed (approved):
Thanks @daledupreez , looks good.

daledupreez commented:
I deployed this to WordPress.com in r251088-wpcom.
